Examples/Io/EDM4hep/src/EDM4hepUtil.cpp

```diff
@@ -0,0 +1,305 @@
+// This file is part of the Acts project.
+//
+// Copyright (C) 2022 CERN for the benefit of the Acts project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#include "ActsExamples/Io/EDM4hep/EDM4hepUtil.hpp"
+
+#include "Acts/Definitions/Units.hpp"
+#include "Acts/EventData/MultiTrajectory.hpp"
+#include "Acts/EventData/MultiTrajectoryHelpers.hpp"
+#include "Acts/Utilities/Helpers.hpp"
+#include "ActsExamples/Digitization/MeasurementCreation.hpp"
+#include "ActsExamples/EventData/Index.hpp"
+#include "ActsExamples/EventData/IndexSourceLink.hpp"
+#include "ActsExamples/EventData/SimHit.hpp"
+#include "ActsExamples/Validation/TrackClassification.hpp"
+
+#include "edm4hep/TrackState.h"
+
+namespace ActsExamples {
+
+ActsFatras::Particle EDM4hepUtil::readParticle(
+    edm4hep::MCParticle from, MapParticleIdFrom particleMapper) {
+  ActsFatras::Barcode particleId = particleMapper(from);
+
+  ActsFatras::Particle to(particleId, (Acts::PdgParticle)from.getPDG(),
+                          from.getCharge() * Acts::UnitConstants::e,
+                          from.getMass() * Acts::UnitConstants::GeV);
+
+  // TODO do we have that in EDM4hep?
+  // particle.setProcess(static_cast<ActsFatras::ProcessType>(data.process));
+
+  to.setPosition4(from.getVertex()[0] * Acts::UnitConstants::mm,
+                  from.getVertex()[1] * Acts::UnitConstants::mm,
+                  from.getVertex()[2] * Acts::UnitConstants::mm,
+                  from.getTime() * Acts::UnitConstants::ns);
+
+  // Only used for direction; normalization/units do not matter
+  to.setDirection(from.getMomentum()[0], from.getMomentum()[1],
+                  from.getMomentum()[2]);
+
+  to.setAbsoluteMomentum(std::hypot(from.getMomentum()[0],
+                                    from.getMomentum()[1],
+                                    from.getMomentum()[2]) *
+                         Acts::UnitConstants::GeV);
+
+  return to;
+}
+
+void EDM4hepUtil::writeParticle(const ActsFatras::Particle& from,
+                                edm4hep::MutableMCParticle to) {
+  // TODO what about particleId?
+
+  to.setPDG(from.pdg());
+  to.setCharge(from.charge() / Acts::UnitConstants::e);
+  to.setMass(from.mass() / Acts::UnitConstants::GeV);
+  to.setVertex({from.position().x(), from.position().y(), from.position().z()});
+  to.setMomentum({(float)from.fourMomentum().x(),
+                  (float)from.fourMomentum().y(),
+                  (float)from.fourMomentum().z()});
+}
+
+ActsFatras::Hit EDM4hepUtil::readSimHit(const edm4hep::SimTrackerHit& from,
+                                        MapParticleIdFrom particleMapper,
+                                        MapGeometryIdFrom geometryMapper) {
+  ActsFatras::Barcode particleId = particleMapper(from.getMCParticle());
+
+  // TODO what about the digitization?
```

Comment:
Can you elaborate?

Reply:
I mixed things up here. simhits are pre-digitization right?

Reply:
They are